src/types.lisp

```diff
@@ -667,28 +680,30 @@ The foreign array must be freed with foreign-array-free."
           (current-offset 0)
           (max-align 1)
           (firstp t))
-      ;; determine offsets
-      (dolist (slotdef slots)
-        (destructuring-bind (slotname type &key (count 1) offset) slotdef
-          (when (eq (canonicalize-foreign-type type) :void)
-            (error "void type not allowed in structure definition: ~S" slotdef))
-          (setq current-offset
-                (or offset
-                    (adjust-for-alignment type current-offset :normal firstp)))
-          (let* ((slot (make-struct-slot slotname current-offset type count))
-                 (align (get-alignment (slot-type slot) :normal firstp)))
-            (setf (gethash slotname (slots struct)) slot)
-            (when (> align max-align)
-              (setq max-align align)))
-          (incf current-offset (* count (foreign-type-size type))))
-        (setq firstp nil))
-      ;; calculate padding and alignment
-      (setf (alignment struct) max-align) ; See point 1 above.
-      (let ((tail-padding (- max-align (rem current-offset max-align))))
-        (unless (= tail-padding max-align) ; See point 3 above.
-          (incf current-offset tail-padding)))
-      (setf (size struct) (or size current-offset))
-      (notice-foreign-type name struct :struct))))
+      (with-tenative-type-definition (name struct :struct)
```

Comment:
the user-level issue is stuff like this:

```
(defcstruct linked-list
  (next (:struct linked-list)))
```

before the patch self-referential type lookup fails while defining the struct...

a simple fix is to introduce a stub for the struct type, and in case any errors happen, then just delete that stub from an UWP.